### ISSUE TITLE
Add option to load clang-tidy plugins

### ISSUE THREAD

**Issue by philnik777** [clangd/clangd#1458]
libc++ has started using the clang-tidy option `--load` and a shared library to be able to run libc++-specific clang-tidy checks. It would be great if one could load the library to get warnings through the language server and being able to apply fixit-hints.

**Comment by HighCommander4** [comment]
Some existing discussion of implementation challenges around loading plugins can be found at https://reviews.llvm.org/D92155

**Comment by kadircet** [comment]
which clang-tidy checks are you talking about? are they not part of default clang-tidy checks in LLVM for some reason?

**Comment by philnik777** [comment]
No, they are clang-tidy checks specifically for libc++. They are implemented [here](https://github.com/llvm/llvm-project/tree/main/libcxx/test/tools/clang_tidy_checks). We don't want them in clang-tidy, since they are very project-specific and having them upstream gives us less flexibility.

**Comment by kadircet** [comment]
I see, as explained in the link provided by @HighCommander4, having plugins/shared-libraries in clangd are really hard, some of the arguments actually goes for your way of using it in a system installed clang-tidy. e.g. there are no guarantees that a new release of clang-tidy will keep the interfaces you're using stable, hence loading your plugin might actually result in crashes, unless it's version locked.

> We don't want them in clang-tidy, since they are very project-specific and having them upstream gives us less flexibility.

They're still part of upstream and they wouldn't be the first project-specific checks in clang-tidy either (see https://github.com/llvm/llvm-project/tree/main/clang-tools-extra/clang-tidy/android/AndroidTidyModule.cpp, and bunch of others like abseil, fuchsia, boost, linuxkernel, ...)

The best course of action I can see for these are, actually having them as a regular module in clang-tidy, that way both clang-tidy CLI and other clang-tools like clangd would be able to report these findings. You can still keep the plugin setup you have in your configuration as well and load it dynamically (but you might want to have a different check name).

Closing this as we're not planning to add support for loading plugins, and I don't think it's something we should be prioritizing any time soon.

**Comment by i-ky** [comment]
> They're still part of upstream and they wouldn't be the first project-specific checks in clang-tidy either

> The best course of action I can see for these are, actually having them as a regular module in clang-tidy, that way both clang-tidy CLI and other clang-tools like clangd would be able to report these findings.

This may be applicable to libc++ specific clang-tidy check, but it is not applicable to out-of-tree clang-tidy plugins in general. There are various reasons why clang-tidy check can't be upstreamed (its application is too niche, legal issues for closed source projects, etc.), so not being able to use out-of-tree clang-tidy plugins with clangd undermines the whole idea of out-of-tree plugins for clang-tidy.

> loading your plugin might actually result in crashes, unless it's version locked.

Regarding possible crashes, clangd's troubleshooting instructions [already have instructions on turning clang-tidy off](https://clangd.llvm.org/troubleshooting#crashes-when-opening-a-file) to separate clang-tidy crashes into their own category. I admit, that it may be a problem. However, it seems that it should be solved on LLVM side (validating plugin's expectations of API/ABI when loading).

I wasn't aware that clang-tidy can do out-of-tree plugins now and at first I was excited to try it out, feeling hopeful about the outcome of this discussion, but categorical "no" from clangd maintainers make clang-tidy plugins less appealing.

**Comment by kadircet** [comment]
> This may be applicable to libc++ specific clang-tidy check, but it is not applicable to out-of-tree clang-tidy plugins in general. There are various reasons why clang-tidy check can't be upstreamed (its application is too niche, legal issues for closed source projects, etc.), so not being able to use out-of-tree clang-tidy plugins with clangd undermines the whole idea of out-of-tree plugins for clang-tidy.

I agree that it would be a nice addition, but as mentioned it comes with a lot of complications and we don't really have any signals indicating this is an important use case to justify eating that complexity.

> Regarding possible crashes, clangd's troubleshooting instructions [already have instructions on turning clang-tidy off](https://clangd.llvm.org/troubleshooting#crashes-when-opening-a-file) to separate clang-tidy crashes into their own category. I admit, that it may be a problem. However, it seems that it should be solved on LLVM side (validating plugin's expectations of API/ABI when loading).

There is no such thing as just clang-tidy plugins, `-load` will imply loading "all clang plugins" (even if that option we have is changed to disable loading clang-tidy plugins at startup somehow). There are projects that use these clang-plugins and things work for them because they have a toolchain they control and both the plugin and clang are version-locked, but developers acquire clangd from different sources most of the time. Hence we had this version discrepancy (and also problems around the way clangd runs compiles, which is different than clang) issues in the past and had to turn off loading plugins for good.

**Comment by i-ky** [comment]
> There are projects that use these clang-plugins and things work for them because they have a toolchain they control and both the plugin and clang are version-locked, but developers acquire clangd from different sources most of the time. Hence we had this version discrepancy (and also problems around the way clangd runs compiles, which is different than clang) issues in the past and had to turn off loading plugins for good.

These projects go to great lengths implementing their own plugin(s), adding a version-locked clangd to the toolchain is a relatively small extra step that won't be a problem. I believe it is safe to assume that _developers using plugins_ and _developers "aquiring clangd from different sources"_ are (almost) disjoint sets, so it can be expected that "discrepancy issues" will be rare enough. Projects using plugins can also be expected to have high level of clang-specific expertise in-house, this will also reduce the rate of false positives reaching clangd maintaners.

**Comment by HighCommander4** [comment]
> These projects go to great lengths implementing their own plugin(s), adding a version-locked clangd to the toolchain is a relatively small extra step that won't be a problem.

Couldn't such a project use a downstream of clangd that incorporates the new check at build time?

**Comment by i-ky** [comment]
> > These projects go to great lengths implementing their own plugin(s), adding a version-locked clangd to the toolchain is a relatively small extra step that won't be a problem.
> 
> Couldn't such a project use a downstream of clangd that incorporates the new check at build time?

Technically, yes. Same can be said about clang-tidy and clang plugins, they can all "live" in LLVM repo forks. According to you, there was no need for clang-tidy and clang plugin infrastructure.

I can also imagine that developing a plugin inside a small independent repo is much simpler than doing so inside a fork of the huge LLVM monorepo.

**Comment by tom-github-account** [comment]
I'd like to vote for this issue.

I develop clang-tidy checks for a very large codebase.  We already build LLVM and make it centrally available for our developers.  We change/extend our clang-tidy checks much more frequently than we update our compiler, so we develop our checks as a stand-alone module.  We effectively already have "a version-locked clangd [in] the toolchain".

In short, I fit exactly into the _ developers using plugins_ category that @i-ky described.

**Comment by ggladilov** [comment]
I'd like to vote for this issue as well.

Our project also has llvm-project as a source dependency, but for another subproject (MLIR). As we extend MLIR in-tree and upgrade LLVM release based on MLIR (which doesn't seem to follow LLVM release schedule yet), it'd add extra complexity for validation to check every time clang-tidy on new revision is still working probably as well. Clang/clang-tidy we use to compile/analyze our code and MLIR we use as a framework are decoupled and don't need to be of the same version.

At the same time, some of our designs rely on static analysis as a way of reflection in C++ (e.g. check particular classes/types don't have non-static data-member variables of pointer/reference types), these checks can't be upstreamed due to being specific to our code base. At the same time, usage of MLIR makes it difficult to keep our checks in-tree of clang-tidy.

**Comment by ggladilov** [comment]
Just to clarify, my main interest here is to pass --load CLI argument to clang-tidy, not clangd. I'm completely out of context how these things are implemented, but maybe just exposing a way to pass CLI arguments to clang-tidy being used by clangd would help to solve the issue?

**Comment by HighCommander4** [comment]
I think that, as with #45, the project would be open to contributions towards plugin support that thoughtfully address the issues raised in the previous technical discussion (mostly in https://reviews.llvm.org/D92155). As such, I think the appropriate status for this issue is "open", and I'm going to take the liberty of reopening it.

**Comment by ggladilov** [comment]
> Just to clarify, my main interest here is to pass --load CLI argument to clang-tidy, not clangd. I'm completely out of context how these things are implemented, but maybe just exposing a way to pass CLI arguments to clang-tidy being used by clangd would help to solve the issue?

To correct myself here: probably exposing CLI of clang-tidy from clangd won't be enough for my use case :(
The idea was to use clangd with clang-tidy inside and pass --load argument to clang-tidy to run custom checks. However, per [@HighCommander4's reply](https://github.com/clangd/clangd/discussions/2280#discussioncomment-11722074):

> Clangd does not use a clang-tidy executable; it links the libraries containing the clang-tidy check implementations into the clangd binary itself.

And from clang-tidy [documentation](https://clang.llvm.org/extra/clang-tidy/Contributing.html#out-of-tree-check-plugins):

> There is no expectations regarding ABI and API stability, so the plugin must be compiled against the version of clang-tidy that will be loading the plugin.

So, as an author of clang-tidy out-of-tree check-plugin I need to know which clang-tidy version clangd is linked against, get access to the same version and link my plugin against it as well. AFAIK, there's no way currently to know the version or if it can be some custom one, not available externally.

**Comment by HighCommander4** [comment]
> So, as an author of clang-tidy out-of-tree check-plugin I need to know which clang-tidy version clangd is linked against, get access to the same version and link my plugin against it as well. AFAIK, there's no way currently to know the version or if it can be some custom one, not available externally.

On this point at least, the answer is simple: clangd and clang-tidy both live in-tree in the same repository (https://github.com/llvm/llvm-project/), and the clang-tidy libraries that clangd links to are built from the same source tree and revision as the rest of clangd, i.e. the versions will be identical (e.g. if you're using clangd 18.1.5, it will link to clang-tidy 18.1.5 libraries).

This doesn't help you in the immediate future, but it will become a consideration if the plugin support being discussed in this issue is eventually added.

**Comment by zhaob1n** [comment]
Another use case here, I've just implemented a clang plugin that provides custom attribute, but the clangd complains with "Unknown attribute".
